tests/test_salcommander_minimal_controllers.py

```diff
@@ -0,0 +1,145 @@
+            initial_log = "initial logLevel.level "
+        else:
+            raise ValueError(f"Argument {language} must be one of 'cpp' or 'java'")
+        return commander, language_upper, initial_log
+
+    def test_cpp_controller(self) -> None:
+        for language in languages:
+            with self.subTest(language=language):
+                self.check_minimal_controller(language, "minimal_cpp_controller.sh")
+
+    def test_java_controller(self) -> None:
+        for language in languages:
+            with self.subTest(language=language):
+                self.check_minimal_controller(language, "minimal_java_controller.sh")
+
+    def test_salobj_controller(self) -> None:
+        for language in languages:
+            with self.subTest(language=language):
+                self.check_minimal_controller(language, "minimal_salobj_controller.py")
+
+    def check_minimal_controller(self, language: str, exec_name: str) -> None:
+        # Start the Commander before the Controller subprocess
+        # to be sure the Commander sees telemetry from the Controller
+        # (telemetry is volatile, so has no historical data).
+        commander, language_upper, initial_log = self.getLanguageVariables(language)
+        commander_path = pathlib.Path(__file__).home() / "repos/ts_sal/bin" / commander
+        print(f"{language_upper} Commander: start {commander_path} with index={self.index} in a subprocess")
+        assert commander_path.is_file()
+        commander_process = subprocess.Popen(
+            [
+                commander_path,
+                str(self.index),
+                str(INITIAL_LOG_LEVEL),
+            ],
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+        )
+        controller_path = (
+            pathlib.Path(__file__).parent.absolute() / "controllers" / exec_name
+        )
+        assert controller_path.is_file()
+        print(f"{language_upper} Controller: start {controller_path} with index={self.index} in a subprocess")
+        controller_process = subprocess.Popen(
+            [str(controller_path), str(self.index), str(INITIAL_LOG_LEVEL)],
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+        )
+
+        try:
+            print(f"{language_upper} Commander: wait for completion")
+            commander_process.wait()
+            commander_output = commander_process.communicate()[0]
+            commander_output = commander_output.decode("utf-8")
+            assert commander_process.returncode == 0
+            print(f"{language_upper} Controller: wait for completion")
+            controller_process.wait()
+            controller_output = controller_process.communicate()[0]
+            controller_output = controller_output.decode("utf-8")
+            print(f"{language_upper} Commander: read initial logLevel")
+            self.assertIn(
+                str(initial_log) + str(INITIAL_LOG_LEVEL), commander_output
```

Comment:
In a future update we should consider making all commanders write identical strings. That would simplify this test since you would not have to keep track of a language-specific "initial_log" variable.

Reply:
I agree it would be better to make these consistent, but I'm not sure it's worth the effort.